### PR TITLE
fix: Parse/Stringify JSON in a locale-independent way

### LINKED ISSUE #407: JSON Module Broken for Locales
**Description**

I have observed that we get JSON values for the duration that are serialized with a comma instead of a dot. So for instance `42,23` instead of `42.23`. This of course breaks the entire payload on the server. Looking at the code a bit it seems like we use `snprintf` which is locale dependent. Likewise for parsing JSON we use `strtod` which again is likely to be broken.

**When does the problem happen**

Observed with sessions, likely also causing other breakage.

**Environment**

- OS: unclear, can't tell from the reports. Most likely German locales on at least linux and windows.

**Steps To Reproduce**

Pseudocode:

```c
#include <locale.h>

setlocale(LC_NUMERIC, "de_DE");
sentry_options_t *options = sentry_options_new();
sentry_options_set_auto_session_tracking(options, true);
sentry_init(options);
```

**Relay observed session**

See the broken `duration` key:

```json
{
  "sid":"SNIP",
  "status":"exited",
  "did":"SNIP",
  "errors":3,
  "started":"2020-10-26T09:43:41.557Z",
  "duration":26033,6,
  "attrs":{"release":"SNIP","environment":null}
}
```

### PR DESCRIPTION
**Pull request by Swatinem**
fixes #407

### CODE REVIEW

File: src/sentry_utils.c [hunk 1]
```diff
@@ -409,3 +422,54 @@ sentry__iso8601_to_msec(const char *iso)
...
 }
+
+#ifdef SENTRY_PLATFORM_WINDOWS
+#    define sentry__locale_t _locale_t
+#else
+#    define sentry__locale_t locale_t
+#endif
+
+static sentry__locale_t
+sentry__c_locale()
```
Comment (mitsuhiko) [inline src/sentry_utils.c:433]: For static, non exported functions I would not use the `sentry__` prefix. This implies the function exists outside of this module which it doesn't.

File: src/sentry_utils.h [hunk 2]
```diff
@@ -165,4 +165,11 @@ char *sentry__msec_time_to_iso8601(uint64_t time);
  */
 uint64_t sentry__iso8601_to_msec(const char *iso);
 
+/**
+ * Locale independent (or rather, using "C" locale) `strtod`.
+ */
+double sentry__strtod_c(const char *ptr, char **endptr);
+
+int sentry__snprintf_c(char *buf, size_t buf_size, const char *fmt, ...);
```
Comment (mitsuhiko) [inline src/sentry_utils.h:173]: Probably should have a similar comment explaining the rationale here.

File: src/sentry_utils.c [hunk 3]
```diff
@@ -409,3 +422,54 @@ sentry__iso8601_to_msec(const char *iso)
...
+    va_list args;
+    va_start(args, fmt);
+
+    int rv;
+#ifdef SENTRY_PLATFORM_WINDOWS
+    rv = _vsnprintf_l(buf, buf_size, fmt, sentry__c_locale(), args);
+#elif defined(SENTRY_PLATFORM_MACOS)
+    rv = vsnprintf_l(buf, buf_size, sentry__c_locale(), fmt, args);
+#else
+    rv = stbsp_vsnprintf(buf, buf_size, fmt, args);
```
Comment (mitsuhiko) [inline src/sentry_utils.c:470]: So I think this is fine, but in theory we could alternatively make a branch here that uses `uselocale` to switch the locale for the current thread and restore it. In the grand scheme of stupid things, that might be an alternative.

File: src/sentry_utils.c [hunk 4]
```diff
@@ -409,3 +422,54 @@ sentry__iso8601_to_msec(const char *iso)
...
+sentry__c_locale()
+{
+    static bool c_locale_initialized = false;
+    static sentry__locale_t c_locale;
+    if (!c_locale_initialized) {
+        c_locale_initialized = true;
+#ifdef SENTRY_PLATFORM_WINDOWS
+        c_locale = _create_locale(LC_ALL, "C");
+#else
+        c_locale = newlocale(LC_ALL_MASK, "C", NULL);
```
Comment (mitsuhiko) [inline src/sentry_utils.c:442]: Not that it matters much, but the third argument is an integer not a pointer. So should probably be `0`.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

File: src/sentry_utils.c (hunk 3 above)
Comment (Swatinem) [reply src/sentry_utils.c:470]: not sure I want to mess with that, I kept it as is for now

Comment (mitsuhiko) [reply src/sentry_utils.c:470]: I think it's fine keeping it like this. We could evaluate it later if we want to trim down on bloat.

**Review by mitsuhiko — APPROVED** [review]

**Pull request merged**